File: packages/doaya_core/lib/src/sync.dart
/// Device ⇄ server sync: the engine and the two interfaces it drives.
///
/// The server (backend/app/sync.py) keeps the latest version of every row;
/// devices push their local changes (an outbox) and pull everything after a
/// cursor. Ledgers are append-only and merge freely; master data is
/// last-writer-wins on (changedAt, deviceId).
library;

/// Tables whose rows never change once written. Must match the server.
const appendOnlyTables = {
  'stock_events',
  'sales',
  'sale_lines',
  'debt_events',
  'returns',
  'return_lines',
  'till_events',
  'purchases',
  'purchase_lines',
  'supplier_debt_events',
  'supplier_returns',
  'expense_events',
  'stocktake_counts',
};

/// Master data: edited in place, last writer wins, may be deleted.
const mutableTables = {
  'devices',
  'settings',
  'employees',
  'products',
  'product_barcodes',
  'customers',
  'suppliers',
  'stocktakes',
  'purchase_orders',
  'purchase_order_lines',
};

/// A local change waiting to be pushed.
class OutgoingChange {
  const OutgoingChange({
    required this.table,
    required this.id,
    required this.changedAt,
    this.data,
    this.deleted = false,
  });

  final String table;
  final String id;

  /// UTC time of the change (ISO-8601); decides last-writer-wins.
  final DateTime changedAt;

  /// The row as column → value; null when deleted.
  final Map<String, Object?>? data;
  final bool deleted;

  String get key => '$table:$id';

  Map<String, Object?> toJson() => {
    'table': table,
    'id': id,
    'changed_at': changedAt.toUtc().toIso8601String(),
    if (!deleted) 'data': data,
    'deleted': deleted,
  };
}

/// An outbox entry: the change plus the local outbox ids it covers.
class PendingChange {
  const PendingChange(this.change, this.outboxIds);
  final OutgoingChange change;
  final List<int> outboxIds;
}

/// A change made on another device.
class IncomingChange {
  const IncomingChange({
    required this.table,
    required this.id,
    required this.changedAt,
    required this.deviceId,
    required this.seq,
    this.data,
    this.deleted = false,
  });

  factory IncomingChange.fromJson(Map<String, Object?> j) => IncomingChange(
    table: j['table']! as String,
    id: j['id']! as String,
    data: (j['data'] as Map<String, Object?>?),
    deleted: j['deleted'] as bool? ?? false,
    changedAt: DateTime.parse(j['changed_at']! as String).toUtc(),
    deviceId: j['device_id']! as String,
    seq: j['seq']! as int,
  );

  final String table;
  final String id;
  final Map<String, Object?>? data;
  final bool deleted;
  final DateTime changedAt;
  final String deviceId;
  final int seq;
}

class PushResult {
  const PushResult({this.accepted = const {}, this.rejected = const {}, this.conflicts = const {}});

  factory PushResult.fromJson(Map<String, Object?> j) => PushResult(
    accepted: {...(j['accepted'] as List).cast<String>()},
    rejected: {
      for (final r in (j['rejected'] as List).cast<Map<String, Object?>>())
        '${r['table']}:${r['id']}',
    },
    conflicts: {...(j['conflicts'] as List).cast<String>()},
  );

  /// Keys (`table:id`) stored or superseded: done on the device.
  final Set<String> accepted;

  /// Keys the server can never take (unknown table, deleting a ledger row).
  final Set<String> rejected;

  /// Ledger keys that already existed with a different body.
  final Set<String> conflicts;
}

class PullPage {
  const PullPage({
    required this.changes,
    required this.cursor,
    required this.more,
    required this.latest,
  });

  factory PullPage.fromJson(Map<String, Object?> j) => PullPage(
    changes: [
      for (final c in (j['changes'] as List).cast<Map<String, Object?>>())
        IncomingChange.fromJson(c),
    ],
    cursor: j['cursor']! as int,
    more: j['more']! as bool,
    latest: j['latest']! as int,
  );

  final List<IncomingChange> changes;
  final int cursor;
  final bool more;
  final int latest;
}

/// Last-writer-wins order: does a change at ([aAt], [aDevice]) beat one at
/// ([bAt], [bDevice])? Later time wins; the device id breaks exact ties.
/// The same rule as the server.
bool newerThan(DateTime aAt, String aDevice, DateTime bAt, String bDevice) {
  final c = aAt.compareTo(bAt);
  return c != 0 ? c > 0 : aDevice.compareTo(bDevice) > 0;
}

/// Thrown by transports when the server can't be reached.
class SyncNetworkException implements Exception {
  const SyncNetworkException(this.message);
  final String message;
  @override
  String toString() => 'SyncNetworkException: $message';
}

/// The network side (HTTP in the app, in-memory in tests).
abstract interface class SyncTransport {
  Future<PushResult> push(List<OutgoingChange> changes);
  Future<PullPage> pull({required int after, required int limit});
}

/// The device side (drift in the app, in-memory in tests).
abstract interface class SyncStore {
  /// Up to [limit] local changes not yet pushed, oldest first, one per row.
  Future<List<PendingChange>> pending(int limit);

  /// Forgets outbox entries the server has handled.
  Future<void> acknowledge(Iterable<int> outboxIds);

  Future<int> cursor();

  /// Applies other devices' changes and saves [cursor], all or nothing.
  Future<void> apply(List<IncomingChange> changes, int cursor);
}

class SyncProgress {
  const SyncProgress({
    required this.pushed,
    required this.pulled,
    required this.cursor,
    required this.latest,
  });
  final int pushed;
  final int pulled;
  final int cursor;

  /// Newest change on the server; cursor/latest is the download progress.
  final int latest;
}

class SyncReport {
  const SyncReport({
    required this.pushed,
    required this.pulled,
    required this.rejected,
    required this.conflicts,
  });
  final int pushed;
  final int pulled;
  final int rejected;
  final int conflicts;
}

/// One sync run: push the whole outbox, then pull until up to date.
///
/// Safe to interrupt anywhere: outbox entries are only forgotten after the
/// server answered, and each pulled page is applied together with its
/// cursor. The next run carries on from there.
class SyncEngine {
  SyncEngine(this.store, this.transport, {this.pushBatch = 500, this.pullLimit = 500});

  final SyncStore store;
  final SyncTransport transport;
  final int pushBatch;
  final int pullLimit;

  Future<SyncReport> sync({void Function(SyncProgress)? onProgress}) async {
    var pushed = 0, pulled = 0, rejected = 0, conflicts = 0;
    var cursor = await store.cursor();

    while (true) {
      final batch = await store.pending(pushBatch);
      if (batch.isEmpty) break;
      final result = await transport.push([for (final p in batch) p.change]);
      final done = <int>[];
      for (final p in batch) {
        final k = p.change.key;
        if (result.accepted.contains(k) || result.rejected.contains(k)) {
          done.addAll(p.outboxIds);
        }
      }
      if (done.isEmpty) {
        throw StateError('server accepted none of ${batch.length} changes');
      }
      await store.acknowledge(done);
      pushed += batch.where((p) => result.accepted.contains(p.change.key)).length;
      rejected += batch.where((p) => result.rejected.contains(p.change.key)).length;
      conflicts += result.conflicts.length;
      onProgress?.call(
        SyncProgress(pushed: pushed, pulled: pulled, cursor: cursor, latest: cursor),
      );
    }

    while (true) {
      final page = await transport.pull(after: cursor, limit: pullLimit);
      await store.apply(page.changes, page.cursor);
      cursor = page.cursor;
      pulled += page.changes.length;
      onProgress?.call(
        SyncProgress(pushed: pushed, pulled: pulled, cursor: cursor, latest: page.latest),
      );
      if (!page.more) break;
    }
    return SyncReport(pushed: pushed, pulled: pulled, rejected: rejected, conflicts: conflicts);
  }
}

File: backend/app/sync.py
"""Push / pull sync for a pharmacy's devices.

Every change a device pushes lands in `sync_rows` with a number from one
sequence. Devices pull "everything after N" from the other devices.

* Ledger tables are append-only: a row is stored once; sending it again is
  harmless (idempotent). A different body for an existing id is ignored and
  reported as a conflict (the first version stays).
* Master data (products, customers…) is last-writer-wins on
  (changed_at, device_id), decided inside the one INSERT … ON CONFLICT.
  Deletions are tombstones (`deleted`, no data) so other devices learn them.
* Pushes of one pharmacy are serialised by a transaction-level advisory lock,
  so sequence numbers commit in order and a pull can never skip a change that
  commits late.
"""

from datetime import datetime

from fastapi import APIRouter, Query
from pydantic import BaseModel, Field
from sqlalchemy import func, select, text
from sqlalchemy.dialects.postgresql import insert

from .deps import Caller, DbSession
from .models import SyncRow

router = APIRouter(prefix="/sync")

# Tables whose rows never change once written (the ledgers).
APPEND_ONLY = frozenset(
    {
        "stock_events",
        "sales",
        "sale_lines",
        "debt_events",
        "returns",
        "return_lines",
        "till_events",
        "purchases",
        "purchase_lines",
        "supplier_debt_events",
        "supplier_returns",
        "expense_events",
        "stocktake_counts",
    }
)

# Master data: edited in place, last writer wins, may be deleted.
MUTABLE = frozenset(
    {
        "devices",
        "settings",
        "employees",
        "products",
        "product_barcodes",
        "customers",
        "suppliers",
        "stocktakes",
        "purchase_orders",
        "purchase_order_lines",
    }
)

MAX_PUSH = 1000
MAX_PULL = 1000


class Change(BaseModel):
    table: str = Field(max_length=60)
    id: str = Field(min_length=1, max_length=80)
    # Ledger rows: when it happened; master rows: when it was last edited.
    changed_at: datetime
    data: dict | None = None
    deleted: bool = False


class PushIn(BaseModel):
    changes: list[Change] = Field(max_length=MAX_PUSH)


class Rejected(BaseModel):
    table: str
    id: str
    reason: str


class PushOut(BaseModel):
    # Everything the device may mark as synced (stored now, already there,
    # or superseded by a newer edit from another device).
    accepted: list[str]
    rejected: list[Rejected]
    # Ledger ids that already existed with a different body.
    conflicts: list[str]


class PulledChange(BaseModel):
    table: str
    id: str
    data: dict | None
    deleted: bool
    changed_at: datetime
    device_id: str
    seq: int


class PullOut(BaseModel):
    changes: list[PulledChange]
    # Pass as `after` next time.
    cursor: int
    more: bool
    # The newest change on the server (for a progress bar on first download).
    latest: int


def _lock_pharmacy(db, pharmacy_id: str) -> None:
    db.execute(text("SELECT pg_advisory_xact_lock(hashtext(:p))"), {"p": pharmacy_id})


@router.post("/push", response_model=PushOut)
def push(body: PushIn, p: Caller, db: DbSession) -> PushOut:
    accepted: list[str] = []
    rejected: list[Rejected] = []
    conflicts: list[str] = []
    _lock_pharmacy(db, p.pharmacy_id)
    for c in body.changes:
        key = f"{c.table}:{c.id}"
        if c.table in APPEND_ONLY:
            if c.deleted or c.data is None:
                rejected.append(Rejected(table=c.table, id=c.id, reason="append_only"))
                continue
            stmt = (
                insert(SyncRow)
                .values(
                    pharmacy_id=p.pharmacy_id,
                    table_name=c.table,
                    row_id=c.id,
                    data=c.data,
                    deleted=False,
                    changed_at=c.changed_at,
                    device_id=p.device_id,
                )
                .on_conflict_do_nothing(constraint="sync_rows_key")
                .returning(SyncRow.id)
            )
            if db.execute(stmt).first() is None:
                existing = db.scalar(
                    select(SyncRow.data).where(
                        SyncRow.pharmacy_id == p.pharmacy_id,
                        SyncRow.table_name == c.table,
                        SyncRow.row_id == c.id,
                    )
                )
                if existing != c.data:
                    conflicts.append(key)
            accepted.append(key)
        elif c.table in MUTABLE:
            if not c.deleted and c.data is None:
                rejected.append(Rejected(table=c.table, id=c.id, reason="no_data"))
                continue
            ins = insert(SyncRow).values(
                pharmacy_id=p.pharmacy_id,
                table_name=c.table,
                row_id=c.id,
                data=None if c.deleted else c.data,
                deleted=c.deleted,
                changed_at=c.changed_at,
                device_id=p.device_id,
            )
            # Newer (changed_at, device_id) wins; the loser is simply dropped.
            stmt = ins.on_conflict_do_update(
                constraint="sync_rows_key",
                set_={
                    "data": ins.excluded.data,
                    "deleted": ins.excluded.deleted,
                    "changed_at": ins.excluded.changed_at,
                    "device_id": ins.excluded.device_id,
                    "seq": func.nextval("change_seq"),
                    "received_at": func.now(),
                },
                where=func.row(SyncRow.changed_at, SyncRow.device_id)
                < func.row(ins.excluded.changed_at, ins.excluded.device_id),
            )
            db.execute(stmt)
            accepted.append(key)
        else:
            rejected.append(Rejected(table=c.table, id=c.id, reason="unknown_table"))
    db.commit()
    return PushOut(accepted=accepted, rejected=rejected, conflicts=conflicts)


@router.get("/pull", response_model=PullOut)
def pull(
    p: Caller,
    db: DbSession,
    after: int = Query(0, ge=0),
    limit: int = Query(500, ge=1, le=MAX_PULL),
) -> PullOut:
    rows = db.scalars(
        select(SyncRow)
        .where(SyncRow.pharmacy_id == p.pharmacy_id, SyncRow.seq > after)
        .order_by(SyncRow.seq)
        .limit(limit)
    ).all()
    latest = db.scalar(
        select(func.coalesce(func.max(SyncRow.seq), 0)).where(SyncRow.pharmacy_id == p.pharmacy_id)
    )
    cursor = rows[-1].seq if rows else max(after, 0)
    return PullOut(
        # A device already has what it wrote itself (and is its last writer).
        changes=[
            PulledChange(
                table=r.table_name,
                id=r.row_id,
                data=r.data,
                deleted=r.deleted,
                changed_at=r.changed_at,
                device_id=r.device_id,
                seq=r.seq,
            )
            for r in rows
            if r.device_id != p.device_id
        ],
        cursor=cursor,
        more=len(rows) == limit,
        latest=latest,
    )
